tests/core/data/test_tftensor.py

```diff
@@ -0,0 +1,171 @@
+import tensorflow as tf
+import pytest
+
+from qutip.core import data
+from qutip.core.data import dense
+from qutip_tensorflow.core.data import TfTensor
+
+from . import conftest
+
+
+# Set up some fixtures for automatic parametrisation.
+
+
+@pytest.fixture(
+    params=[
+        pytest.param((1, 5), id="ket"),
+        pytest.param((5, 1), id="bra"),
+        pytest.param((5, 5), id="square"),
+        pytest.param((2, 4), id="wide"),
+        pytest.param((4, 2), id="tall"),
+    ]
+)
+def shape(request):
+    return request.param
+
+
+@pytest.fixture(params=[True, False], ids=["Fortran", "C"])
+def fortran(request):
+    return request.param
+
+
+def _valid_numpy():
+    # Arbitrary valid numpy array.
+    return conftest.random_numpy_dense((5, 5), False)
+
+
+def _valid_tensor():
+    # Arbitrary valid tensor array.
+    return conftest.random_tensor_dense((5, 5))
+
+
+def _valid_list():
+    # Arbitrary valid tensor array.
+    return conftest.random_numpy_dense((5, 5), False).tolist()
+
+
+@pytest.fixture(scope="function")
+def numpy_dense(shape, fortran):
+    return conftest.random_numpy_dense(shape, fortran)
+
+
+@pytest.fixture(scope="function")
+def tensor_dense(shape):
+    return conftest.random_tensor_dense(shape)
+
+
+@pytest.fixture(scope="function")
+def list_dense(shape):
+    return conftest.random_numpy_dense(shape, fortran).tolist()
+
+
+@pytest.fixture(scope="function")
+def data_tensor_dense(tensor_dense):
+    return TfTensor(tensor_dense)
+
+
+class TestClassMethods:
+    def test_init_from_list(self, list_dense, shape):
+        test = TfTensor(list_dense, shape)
+        assert test.shape == shape
+        assert np.all(test.to_array() == np.array(list_dense))
+
+    def test_init_from_ndarray(self, numpy_dense):
+        test = TfTensor(numpy_dense)
+        assert test.shape == numpy_dense.shape
+        assert np.all(test.to_array() == numpy_dense)
+
+    def test_init_from_tensor(self, tensor_dense):
+        test = TfTensor(tensor_dense)
+        assert test.shape == tuple(tensor_dense.shape.as_list())
+        assert np.all(test.to_array() == tensor_dense)
+
+        # by default we do not return a copy
+        assert test._tf is tensor_dense
+
+    @pytest.mark.parametrize(
+        "dtype", ["complex128", "float64", "int32", "int64", "uint32"]
+    )
+    def test_init_from_list_other_dtype(self, shape, dtype):
+        _numpy_dense = np.random.rand(*shape).astype(dtype, casting="unsafe")
+        _list_dense = _numpy_dense.tolist()
+        test = TfTensor(_list_dense)
+        assert test.shape == shape
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert_almost_equal(test.to_array(), _list_dense)
+
+    @pytest.mark.parametrize(
+        "dtype", ["complex128", "float64", "int32", "int64", "uint32"]
+    )
+    def test_init_from_ndarray_other_dtype(self, shape, dtype):
+        _numpy_dense = np.random.rand(*shape).astype(dtype, casting="unsafe")
```

Comment:
Be careful with this casting; `np.random.rand()` returns a float on [0, 1], which will be cast to a integer 0 every time bar ~once every 2^53 times.  So you're basically just testing on the zero matrix every time, which probably isn't what you meant.

Reply:
I see the problem but given there is such a low chance to occur I think I will let it be for the moment and focus on the next PR.

Reply:
The point is more that the test isn't correct - you're not actually passing it random data, you're just passing a zero matrix. You could basically just fix it by generating random unsigned integers instead of random floats, and then you wouldn't need the unsafe cast either.